Proposal for supporting deep copy of out parameters (#3699)
openenclave/openenclave: docs/DesignDocs/DeepCopyOutParameters.md

Support Deep Copy of Variable-Length Out Parameters
=====

This document provides a brief overview of the oeedger8r's deep copy feature and points out
its missing piece (i.e., the support of `out` parameter). Next, the document discusses two
designs that support the `out` parameter based on the availability of shared memory between
a host and an enclave.

Introduction
------------

The oeedger8r tool supports the deep copy feature that allows a user to pass a struct with nested pointers
in ECALLs and OCALLs. To leverage this feature, the user should define the struct as part of an EDL file
and use proper annotations on both the struct and the ECALL/OCALL, which intends to pass the struct.

> An example of strcut definitions is as follows.

Reviewer: ```suggestion
An example of struct definitions is as follows.
```

  of the `in` parameter, the caller is responsible for initializing `b` and the callee can read
  the content of the `b`. Besides, callee's modifications to `b` will now reflect on the caller
  site with a limitation that the callee cannot re-allocate or resize `b->buf`
  (which could lead to undefined behaviors).

The case that remains unsupported is the `out` parameter as shown in the following example.
```
// EDL definition
void foo[out, count=1] Blob *b);
```
The major difference between the `out` parameter and the other two is that caller is no longer responsible
for initializing `b` (i.e., the caller can simply pass a `b` pointer with `b->buf` equals to `NULL`).
Instead, the callee takes charge of initializing `b`, which implies that `b->buf` is variable-length.
Such support is required by typical scenarios in which the caller cannot determine the size of `b->buf`.

Proposed Design
-----------

An important constraint that affects the design of the `out` parameter support is the availability of
shared memory between a host and an enclave. Given that this support may not be general across TEEs,
we discuss two designs in this section.

> The following code snippet serves as a common example that both two design aims to support.

Reviewer: ```suggestion
The following code snippet serves as a common example that both designs aim to support.
```

- Host
  ```
  Blob b;
  // b->len and b->buf have not been initialized yet.
  foo(enclave, &b);
  // b->len and b->buf have been initialized.
  // b->len == 10 and b->data equals to data.

>   ```

Reviewer: Incomplete code.  How is b->data freed?   Please include in the sample what the developer needs to do to free the blob.  e.g., is it
free(b->data);
?

Author: Updated.

- Enclave
  ```
  // User-defined code
  void foo(Blob *b)
  {
    b->len = 10;
    b->buf = malloc(b->size);
    memcpy(b->buf, data, b->size);
  }
  ```

### Design with Shared Memory

With the support of shared memory, the design is more straightforward.
The core idea of the design is that after the callee initializes the `b`,

> it also allocates a host memory and copies the content `b` to the allocated

Reviewer: Line 29 talks about both directions (ECALL and OCALL) but here only talks about one direction (ECALL).
What about out parameters of OCALLs?

Author: Updated.

memory. Note that the latter part can be done by the oeedger8r-generated code.

- Host

  The host invokes the oeedger8r-generated code stub as usual.
  ```
  // Oeedger8r-generated code
  void foo(oe_enclave* enclave, Blob *b)
  {
    ...
    oe_call_enclave_function(...);
    ...
  }
  ```

- Enclave

  Upon the dispatching of the ECALL, the oeedger8r-generated
  `ecall_foo` is invoked as usual. However, right after the
  invocation of the user-defined `foo`, the design adds a piece
  of code that allocates host memory and copies the content
  of `b` to the allocated memory.

  ```
  // Oeedger8r-generated code
  void ecall_foo()
  {
    ...
    Blob b;
    foo(&b);
    ...
    // Initialize the host memory
    Blob *lhs = (Blob *) pargs->in;
    lhs->len = b.len;
    lhs->buf = oe_host_malloc(lhs->len);
    if (lhs->buf && b.buf)
      memcpy(lhs->buf, b.buf, lhs->len);
    ...
  }

>   ```

Reviewer: Question: will the above also work for out parameters on _switchless_ calls?

Author: Good point. That's something we need to consider. Taking the switchless calls into account may further complicate the design without shared memory.

Reviewer 2: switchless calls are not supported on OPTEE.

### Design without Shared Memory

The design without shared memory is more complicated in terms of implementation.
The high-level idea of the design is that after the callee initializes `b`,

> explicitly invoking an OCALL that takes `b` as an `in` parameter.

Reviewer: Again this only talks about out parameters of ECALLs, what about out parameters of OCALLs?

Author: Updated.

- Host

  In addition to the usual oeedger8r-generated ECALL wrapper, the design
  adds an internal OCALL definition and implementation that will be
  automatically invoked by the enclave.

>   Note that goal of `b_out` being `user_check` is to forward the original value

Reviewer: ```suggestion
  Note that the goal of `b_out` being `user_check` is to forward the original value
```

more information regarding the syntax of oeedger8r.

To pass the `Blob` struct in an ECALL/OCALL, the oeedger8r currently supports the following two cases:
- `in` parameter
  ```
  // EDL definition
  void foo([in, count=1] Blob *b);
  ```

  The above code snippet shows an example of passing `b` as an `in` parameter. In this case,
  the caller is responsible for initializing `b`, including `b->len` and the memory pointed by `b->buf`.
  The callee can then read the content of `b`. However, any callee's modifications to
  `b` will not reflect on the caller site (i.e., `b` is effectively read-only to the callee).

- `inout` parameter
  ```
  // EDL definition
  void foo[in, out, count=1] Blob *b);
  ```

  The above definition is an example of passing `b` as an `inout` parameter. Similar to the case
  of the `in` parameter, the caller is responsible for initializing `b` and the callee can read
  the content of the `b`. Besides, callee's modifications to `b` will now reflect on the caller
  site with a limitation that the callee cannot re-allocate or resize `b->buf`
  (which could lead to undefined behaviors).

> The case that remains unsupported is the `out` parameter as shown in the following example.

Author: Updated the proposal to mention that the design does not change the behavior of existing `inout` parameter.

for initializing `b` (i.e., the caller can simply pass a `b` pointer with `b->buf` equals to `NULL`).
Instead, the callee takes charge of initializing `b`, which implies that `b->buf` is variable-length.
Such support is required by typical scenarios in which the caller cannot determine the size of `b->buf`.

Proposed Design
-----------

An important constraint that affects the design of the `out` parameter support is the availability of
shared memory between a host and an enclave. Given that this support may not be general across TEEs,
we discuss two designs in this section.

The following code snippet serves as a common example that both designs aim to support.
- Caller
  ```
  Blob b;
  // b->len and b->buf have not been initialized yet.
  foo(enclave, &b);
  // b->len and b->buf have been initialized.
  // b->len == 10 and b->data equals to data.
  free(b->buf);
  ```

- Callee
  ```
  // User-defined code
  void foo(Blob *b)
  {
    b->len = 10;
    b->buf = malloc(b->size);
    memcpy(b->buf, data, b->size);
  }
  ```
  Note that `b->buf` also needs to be freed on the callee side, which is left to the implementation

>   details.

Author: We can free the allocated buffer automatically (in the oeedger8r-generated code). However, this may not work with 
 a corner case: instead of allocating memory, the callee assigns a global pointer to the `b->buf`. See the following example.
```
/* Managed somewhere else */
char *g_buf;
size_t g_buf_len;

void foo(Blob *b)
{
    b->len = g_buf_len;
    b->buf = g_buf;
}
```

>   details.

Reviewer: I don't understand what you're proposing as an answer.

>   details.

Author: Now the callee-allocated memory will be freed automatically. Please check the updated document.

### Design with Shared Memory

With the support of shared memory, the design is more straightforward.
The core idea of the design is that after the callee initializes the `b`,
it also allocates a host memory and copies the content `b` to the allocated
memory. Note that the latter part can be done by the oeedger8r-generated code.

- ECALL
  - Host

    The host invokes the oeedger8r-generated ECALL stub as usual.
    ```
    // Oeedger8r-generated code
    void foo(oe_enclave* enclave, Blob *b)
    {
      ...
      oe_call_enclave_function(...);
      ...
    }
    ```

  - Enclave

    Upon the dispatching of the ECALL, the oeedger8r-generated
    `ecall_foo` is invoked as usual. However, right after the
    invocation of the user-defined `foo`, the design adds a piece
    of code that allocates host memory and copies the content
    of `b` to the allocated memory.

    ```
    // Oeedger8r-generated code
    void ecall_foo()
    {
      ...
      Blob b;
      foo(&b);
      ...
      // Initialize the host memory
      Blob *lhs = (Blob *) pargs->in;
      lhs->len = b.len;
      lhs->buf = oe_host_malloc(lhs->len);
      if (lhs->buf && b.buf)

>         memcpy(lhs->buf, b.buf, lhs->len);

Author: Thanks for the suggestion. But it's unclear how memdup supports the copy of nested pointers, which I need to think.

>         memcpy(lhs->buf, b.buf, lhs->len);

Author: The design has been updated such that the memory transmission is now handled internally, not by the oeedger8r-generated code. Please take a look.

      ...
    }
    ```

- OCALL
  - Enclave

    Given that the shared memory is available (i.e., the enclave
    can access host memory), the design of the OCALL adds the similar
    piece of the code on the enclave side that allocates the memory
    and copies the content in the host memory to the allocated memory.

    ```
    // Oeedger8r-generated code
    void foo(Blob *b)
    {
      ...
      oe_call_host_function(...);
      ...
      if (b)
      {
        Blob *rhs = (Blob *) (_output_buffer + _output_buffer_offset);
        b->len = rhs->len;
        b->buf = oe_malloc(b->len);
        if (b->buf && rhs->buf)
          memcpy(b->buf, rhs->buf, b->len);
      }
    }
    ```

  - Host

    In the case of OCALL, the callee invokes the user-defined `foo` as usual.
    ```
    // Oeedger8r-generated code
    void ocall_foo()
    {
      ...
      Blob b;
      foo(&b);

>       ...

Reviewer: why is there no free(b->buf); ?

Author: See my previous comment regarding the special case.

      /* Always free the intermediate enclave buffers. */
      if (_pargs_in->b)
      {
        oe_free(_pargs_in->b->buf);
        _pargs_in->b->buf = NULL;
      }
    }
    ```

- OCALL
  - Enclave

    Given that the shared memory is available (i.e., the enclave
    can access host memory), the design of the OCALL adds the similar
    piece of the code on the enclave side that allocates the memory
    and copies the content in the host memory to the allocated memory.

    ```
    // Oeedger8r-generated code
    void foo(Blob *b)
    {
      ...
      oe_call_host_function(...);
      ...
      /* Initialize the deep-copyable, out-only parameters pointed by _pargs_out to NULL.
         These parameters will be freed at the end of the function if not NULL. */
      pargs_out->b = NULL;

      if (b)
      {
        _pargs_out->b = (Blob *) (_output_buffer + _output_buffer_offset);
        OE_ADD_SIZE(_output_buffer_offset, size(Blob));
        b->len = _pargs_out->b->len;
        b->buf = NULL;
        if (_pargs_out->b->buf)
        {
          b->buf = oe_malloc(b->len);
          if (!b->buf)
          {
            _result = OE_FAILURE;
            goto done;
          }

>           memcpy(b->buf, _pargs_out->b->buf, b->len);

Reviewer 3: If higher level code did not verify that `_pargs_out->b->buf`, with length of `b->len`, is fully outside of the enclave memory already, the code here must verify that before performing the copy.

Author: `b->len` is inside the enclave while `_pargs_out->b->len` is on the hosot.

Author: The design has been modified and hence the comments are no longer applicable.

  void foo([in, count=1] Blob *b);
  ```

  The above code snippet shows an example of passing `b` as an `in` parameter. In this case,
  the caller is responsible for initializing `b`, including `b->len` and the memory pointed by `b->buf`.
  The callee can then read the content of `b`. However, any callee's modifications to
  `b` will not reflect on the caller site (i.e., `b` is effectively read-only to the callee).

- `inout` parameter
  ```
  // EDL definition
  void foo[in, out, count=1] Blob *b);
  ```

  The above definition is an example of passing `b` as an `inout` parameter. Similar to the case
  of the `in` parameter, the caller is responsible for initializing `b` and the callee can read
  the content of the `b`. Besides, callee's modifications to `b` will now reflect on the caller
  site with a limitation that the callee cannot re-allocate or resize `b->buf`
  (which could lead to undefined behaviors).

The case that remains unsupported is the `out` parameter as shown in the following example.
```
// EDL definition
void foo[out, count=1] Blob *b);
```
The major difference between the `out` parameter and the other two is that caller is no longer responsible

> for initializing `b` (i.e., the caller can simply pass a `b` pointer with `b->buf` equals to `NULL`).

Author: Updated the doc.

Instead, the callee takes charge of initializing `b`, which implies that `b->buf` is variable-length.
Such support is required by typical scenarios in which the caller cannot determine the size of `b->buf`.

Proposed Design
-----------

An important constraint that affects the design of the `out` parameter support is the availability of
shared memory between a host and an enclave. Given that this support may not be general across TEEs,
we discuss two designs in this section.

The following code snippet serves as a common example that both designs aim to support.
- Caller
  ```
  Blob b;
  // b->len and b->buf have not been initialized yet.
  foo(enclave, &b);
  // b->len and b->buf have been initialized.
  // b->len == 10 and b->data equals to data.

>   free(b->buf);

Author: Updated the example.

  ```

- Callee
  ```
  // User-defined code
  void foo(Blob *b)
  {
    int result = 0;
    b->len = 10;

>     b->buf = malloc(b->size);

Author: Updated design.

follows.

Note that this design is compatible with switchelss call feature, which also requires
shared memory.

- ECALL
  - Host

    The host invokes the oeedger8r-generated ECALL stub as usual.
    ```
    // Oeedger8r-generated code
    void foo(oe_enclave* enclave, Blob *b)
    {
      ...
      oe_call_enclave_function(...);
      ...
    }
    ```

  - Enclave

    Upon the dispatching of the ECALL, the oeedger8r-generated
    `ecall_foo` is invoked as usual. However, right after the
    invocation of the user-defined `foo`, the design adds a piece
    of code that allocates host memory and copies the content
    of `b` to the allocated memory.

    ```
    // Oeedger8r-generated code
    void ecall_foo()
    {
      ...
      foo(_pargs_in->b);
      ...
      /* _pargs_out holds the pointer of b on the host. */
      if (_pargs_out->b)
      {
        Blob *_rhs = (Blob *) _pargs_in->b;
        _pargs_out->len = _rhs->len;
        _pargs_out->b = NULL;
        if (_rhs->b)
        {
          /* Use a local variable to prevent the host from
             manipulating the address value used by memcpy below. */

>           char* tmp_buf = oe_host_malloc(_rhs->len);

Author: Added `oe_allocate_host_buffer` and `oe_free_host_buffer`.

          _pargs_out->buf = tmp_buf;
          if (!tmp_buf)
          {
            _result = OE_FAILURE;
            goto done;
          }
          memcpy(tmp_buf, _rhs->buf, _rhs->len);
        }
      }
      ...
    done:
      /* Free host buffers on failure. */
      if (_result != OE_OK)
      {
        if (_pargs_out->b)
        {
          oe_host_free(_pargs_out->b->buf);
          _pargs_out->b->buf = NULL;
        }
      }
      /* Always free the intermediate enclave buffers. */
      if (_pargs_in->b)
      {

>         oe_free(_pargs_in->b->buf);

Author: Updated.

Support Deep Copy of Variable-Length Out Parameters
=====

This document provides a brief overview of the oeedger8r's deep copy feature and points out
its missing piece: the support of the deep-copy `out` parameter. Next, the document describes a
design that supports the deep-copy `out` parameter.

Introduction
------------

The oeedger8r tool supports the feature of deep copy, which allows a user to pass a struct with nested pointers

> in ECALLs and OCALLs. To leverage this feature, the user should define the struct as part of an EDL file

Reviewer: Recommend adding a reference to full_edl_serialization.md, and updating full_edl_serialization.md with a reference to this design doc.  Also full_edl_serialization.md includes the statement:

> Although EDL provides sufficient mechanisms for performing full serialization, the generator does not currently enforce these mechanisms. By default, the generator produces edge routines that implicitly copy untrusted memory references into the enclave. This results in potential security flaws and non-portable code.
> 
> To overcome these problems, the generator can be modified to warn when an EDL specification cannot be fully serialized.

I'd suggest updating that text in this PR.

Author: Updated. Please take a look.

and use proper annotations on both the struct and the ECALL/OCALL, which intends to pass the struct.
An example of struct definitions is as follows.

```
struct Blob
{
  size_t len;
  [size=len] char* buf;
};

struct NestedBlob
{
  size_t num;
  [count=num] Blob* blob_array;
};
```

The `num` variable specified to the `count` attribute represents the length (i.e., the number of elements)
of the `blob_array`. Similarly, the `len` variable specified to the `size` attribute represents the size (in bytes) of the `buf`.
Refer to the [Grammar documentation](https://github.com/openenclave/oeedger8r-cpp/blob/master/docs/Grammar.md) for
more information regarding the syntax of oeedger8r.

> To pass the `NestedBlob` struct in an ECALL/OCALL, the oeedger8r currently supports the following two cases:

Reviewer: ```suggestion
To pass the `NestedBlob` struct in an ECALL/OCALL, oeedger8r currently supports the following two cases:
```

  site with a limitation that the callee cannot re-allocate or resize `nb->blob_array` and each `buf`
  in the nested element (which could lead to undefined behaviors).

The case that remains unsupported is the deep-copy `out` parameter as shown in the following example.
```
// EDL definition
void foo([out, count=1] NestedBlob *nb);
```
The major difference between the `out` parameter and the other two is that the caller is no longer responsible
for initializing `nb`.
Instead, the callee takes charge of initializing `nb`, which implies that `b->blob_array` and each of the nested
`buf` are variable-length. Such support is required by typical scenarios in which the caller cannot determine the size of
nested buffers in a struct. Note that the caller is still responsible for allocating the `nb` itself. Please refer to the code snippet
in the following section for an example.

Proposed Design
-----------

This document proposes to add the support of deep-copy `out` parameter to OE SDK.

### User Experience

The following code snippet serves as an example that the proposal aims to support.

- Caller
  

>   The caller, either an enclave or a host, is required to prepare an instant of `NestedBlob`.
>   In this example, `nb` is the instant of `NestedBlob`. The caller then passes `nb` to

Reviewer: ```suggestion
  The caller, either an enclave or a host, is required to prepare an instance of `NestedBlob`.
  In this example, `nb` is the instance of `NestedBlob`. The caller then passes `nb` to
```

    Each nested buffer of `nb` is allocated separately and is therefore required to
    free each of the buffers as shown in the example.

- Callee

  The callee is responsible for initializing the `nb` if the `nb` is not `NULL`.
  The `nb` being `NULL` indicates that the caller passes `NULL` to `foo`.
  
  ```
  // User-defined code
  void foo(NestedBlob *nb)
  {
      if (!nb)
          return;
      nb->num = 5;
      nb->blob_array = (Blob*)malloc(5 * sizeof(Blob));
      for (size_t i = 0; i < 5; i++)
      {
          Blob* b = &ptr->blob_array[i];
          b->len = 10;
          b->buf = (char*)malloc(10);
          memset(b->buf, 'A', 10);
      }
  }
  ```

  - Memory management

    The callee should **allocate each nested buffer separately on the heap** as shown in the example.
    These allocated buffers are freed automatically by the oeedger8r-generated code and therefore the
    callee does not need to take further actions. 
    
Note that the proposal does not affect the behavior of the existing deep-copy `inout` parameter.

### Design

The idea of the design is that after the callee (either a host or an enclave) initializes `nb`,

> the callee serializes the content of `nb` into a local buffer. Next, the content of the buffer

Reviewer: ```suggestion
oeedger8r-generated code serializes the content of `nb` into a local buffer. Next, the content of the buffer
```

is copied over to a caller's buffer. How the buffer is transmitted is TEE-specific;
in the case of SGX in which the enclave has access to the memory on the host, the enclave, either
being the caller or callee, is always responsible for copying the buffer.
Finally, the caller unserializes the received content, which fills up the `nb`.
The rest of the section walks through modified oeedger8r-generated code for both ECALL and OCALL that realizes the design.

> Note that this design is compatible with switchelss call feature, which also requires shared memory.

Reviewer: ```suggestion
Note that this design is compatible with the switchless call feature, which also requires shared memory.
```

- ECALL

>   The design adds `deepcopy_out_buffer` and `deepcopy_out_buffer_size` members to every internal structs of ECALLs.

Reviewer: ```suggestion
  The design adds `deepcopy_out_buffer` and `deepcopy_out_buffer_size` members to every internal struct of ECALLs.
```

  The two members are used to hold and to obtain the enclave-set content by the enclave and the host, correspondingly.
  How the content is transmitted across the boundary is TEE-specific, depending on the availability of the shared memory

>   between the host and the enclave. The value (zero or not) of the two members are also used by the

Reviewer: ```suggestion
  between the host and the enclave. The values (zero or not) of the two members are also used by the
```

  infrastructure to determine whether to transmit the buffer or not.
  
  ```diff
  typedef struct _foo_args_t
  {
      oe_result_t _result;
  +   uint8_t* deepcopy_out_buffer;
  +   size_t deepcopy_out_buffer_size;
      NestedBlob* ptr;
  } foo_args_t;
  ```
  - Host

    Similar to a normal ECALL, the host prepares parameters and invokes `oe_call_enclave_function`.
    After returning from `oe_call_enclave_function`, the host obtains `_deepcopy_out_buffer` and
    `_deepcopy_out_buffer_size` from `pargs_out`. `_deepcopy_out_buffer` holds a local copy of the enclave-set content.
    Next, the host uses the `OE_SET_DEEPCOPY_OUT_PARAM` macro to unserialize the enclave-set content. After unserializing,
    the host validates the `_deepcopy_out_buffer_offset` (increased during unserialization) against `_deepcopy_out_buffer_size`.
    Before finishing the function, the host frees `_deepcopy_out_buffer`.

>     

Reviewer: ```suggestion
    
The following is an example of oeedger8r-generated code:
```

    +   pargs_out->deepcopy_out_buffer_size = _deepcopy_out_buffer_size;

        /* Success. */
        _result = OE_OK;
        *output_bytes_written = output_buffer_offset;

    done:
    +   /* Free pargs_out->deepcopy_out_buffer on failure. */
    +   if (_result != OE_OK)
    +   {
    +       oe_free(pargs_out->deepcopy_out_buffer);
    +       pargs_out->deepcopy_out_buffer = NULL;
    +       pargs_out->deepcopy_out_buffer_size = 0;
    +   }

    +   /* Free nested buffers allocated by the user function. */
    +   if (pargs_in->ptr)
    +   {
    +       for (size_t _i_1 = 0; _i_1 < 1; _i_1++)
    +       {
    +           if (pargs_in->ptr[_i_1].blob_array)
    +           {
    +               for (size_t _i_2 = 0; _i_2 < pargs_in->ptr[_i_1].num; _i_2++)
    +               {
    +                   oe_free(pargs_in->ptr[_i_1].blob_array[_i_2].buf);
    +               }
    +           }
    +           oe_free(pargs_in->ptr[_i_1].blob_array);
    +       }
    +   }

        if (output_buffer_size >= sizeof(*pargs_out) &&
            oe_is_within_enclave(pargs_out, output_buffer_size))
            pargs_out->_result = _result;
    }
    ```

- OCALL

>   Similar to the case of ECALL, the design adds the same modifications to every internal structs of OCALLs.

Reviewer: ```suggestion
  Similar to the case of ECALL, the design adds the same modifications to every internal struct of OCALLs.
```

    +   }

    +   /* Free nested buffers allocated by the host. */
    +   if (pargs_in->ptr)
    +   {
    +       for (size_t _i_1 = 0; _i_1 < 1; _i_1++)
    +       {
    +           if (pargs_in->ptr[_i_1].blob_array)
    +           {
    +               for (size_t _i_2 = 0; _i_2 < pargs_in->ptr[_i_1].num; _i_2++)
    +               {
    +                   free(pargs_in->ptr[_i_1].blob_array[_i_2].buf);
    +                   pargs_in->ptr[_i_1].blob_array[_i_2].buf = NULL;
    +               }
    +           }
    +           free(pargs_in->ptr[_i_1].blob_array);
    +           pargs_in->ptr[_i_1].blob_array = NULL;
    +       }
    +   }

        if (pargs_out && output_buffer_size >= sizeof(*pargs_out))
            pargs_out->_result = _result;
    }
    ```

Conclusion and Plan
-----------

This document presents the design, which supports deep-copy `out` parameter.
Moving forward, we plan to support the design as an experimental feature first.
That is, users need to explicitly opt-in the feature (i.e., specifying `--experimental` flag in the
command line when invoking the oeedger8r tool). Doing so allows OE to support the deep-copy `out` parameter

> in SGX while not introducing breaking changes to the OE's support on the OP-TEE.

Reviewer: ```suggestion
in SGX while not introducing breaking changes to the OE's support on OP-TEE.
```

> in SGX while not introducing breaking changes to the OE's support on the OP-TEE.

Reviewer: ```suggestion
in SGX while not introducing breaking changes to the OE's support on the OP-TEE.
```
@HernanGatta is there anything in here that would break something that works on OP-TEE?  If not, then this doesn't need to be experimental.

Author: The design won't break existing OP-TEE ECALLs/OCALLs that don't use the deep-copy out annotations (there shouldn't be any).
Even if developers try to use deep-copy out on OP-TEE, the compilation would go thru but the code just won't work unless we implement the OP-TEE support internally.

Given that we currently don't specify target TEE to the oeedger8r tool, we don't have a good way to tell developers that deep-copy out is not supported on OP-TEE. As a result, the current plan is guarding the feature with the `--experimental` flag, which makes the feature implicit. If the flag is not set, the developer will see an error message like:
```
`out' for user defined type `MyStruct\\*' requires the --experimental option`.
```

@dthaler @anakrish Do you think we can make this as non-experimental? We do have the switchless call feature that is similar to this case (only works for SGX).

Alternatively, we could introduce a TEE flag to oeedger8r or let the tool to detect the target based on the platform (x86 or arm). We could also add the OP-TEE support, which requires @HernanGatta's inputs. I think both of them could be future work.

Reviewer 4: Let's first merge this in as experimental, and get feedback from the user that wants this.

Reviewer: Unless we know of things that would break OP-TEE, I don't think OP-TEE should be any reason to make this be experimental.  If we think it needs to be experimental for SGX because SGX might break, then ok.  But I'd probably argue for non-experimental.

Author: We're thinking of following the previous path on deep-copy support: making the feature as experimental first and then get feedback from users/ask for security reviews before making this as non-experimental.

Reviewer 2: I agree with what Ming-Wei said. Let’s get feedback and see if some folks can look at the security implications. Cc @mrragava

  }
  ```

  - Memory management

    The callee should **allocate each nested buffer separately on the heap** as shown in the example.
    These allocated buffers are freed automatically by the oeedger8r-generated code and therefore the
    callee does not need to take further actions. 
    
Note that the proposal does not affect the behavior of the existing deep-copy `inout` parameter.

### Design

The design assumes that the shared memory between the host and the enclave is available.
The idea of the design is that after the callee (either a host or an enclave) initializes `nb`,
the callee serializes the content of `nb` into a local buffer. Next, the content of the buffer
is copied over to a caller's buffer. Given that only the enclave has access to the memory
on the host, the enclave, either being the caller or callee, is always responsible for copying the buffer.
Finally, the caller unserializes the received content, which fills up the `nb`.
The rest of the section walks through modified oeedger8r-generated code for both ECALL and OCALL that realizes the design.

Note that this design is compatible with switchelss call feature, which also requires shared memory.

- ECALL

>   The design adds a `_host_buffer` member to the internal struct that holds the shared pointer.

Reviewer 4: You could also add a `_host_buffer_size` member. The first 3 fields of any marshalling struct would then be:
```c
typedef struct _fcn_args_t
  {
      oe_result_t _result;
      uint8_t* _host_buffer;
      size_t _host_buffer_size;
      ... // other marshaling fields
  }
```

Author: Updated.

        /* No return value. */

    +   OE_READ_OUT_PARAM(ptr, (size_t)(((size_t)1 * sizeof(NestedBlob))));
    +   if (ptr)
    +       OE_SET_DEEPCOPY_OUT_PARAM(ptr->blob_array, ((size_t)ptr->num * sizeof(Blob)), Blob*);
    +   for (size_t _i_2 = 0; _i_2 < ptr->num; _i_2++)
    +   {
    +       if (ptr && ptr->blob_array)
    +           OE_SET_DEEPCOPY_OUT_PARAM(ptr->blob_array[_i_2].buf, ptr->blob_array[_i_2].len, char*);
    +   }

    +   if (_deepcopy_out_buffer_offset != _deepcopy_out_buffer_size)
    +   {
    +       _result = OE_FAILURE;
    +       goto done;
    +   }

        _result = OE_OK;

    done:
        if (_buffer)
            free(_buffer);

    +   if (_deepcopy_out_buffer)
    +       free(_deepcopy_out_buffer);

        return _result;
    }
    ```

  - Enclave

    Upon the dispatching of the ECALL, the oeedger8r-generated
    `ecall_foo` is invoked as usual. After the invocation of
    the user-defined `foo` function, the enclave calculates the size
    of content set by `foo`. Next, the enclave allocates an in-enclave
    buffer, `_deepcopy_out_buffer`, and serializes the content into the buffer using the
    `OE_WRITE_DEEPCOPY_OUT_PARAM` macro.

>     After serializing, the enclave allocates a host buffer (held by the `pargs_out->_host_buffer`

Reviewer 4: This does not have to be done by the generated code. 
The generated code just needs to do 
```c
    args->_host_buffer = _deepcopy_out_buffer;
    args->_host_buffer_size = _deepcopy_out_buffer_size;
```

The ECALL dispatching infrastructure can determine the best way to marshal _host_buffer over to the host, according to the platform (SGX/OPTEE).

Author: Updated.

    +       OE_ADD_SIZE(_deepcopy_out_buffer_size, pargs_in->ptr->blob_array[_i_2].len);
    +   }

    +   if (_deepcopy_out_buffer_size)
    +   {
    +       _deepcopy_out_buffer = (uint8_t*) malloc(_deepcopy_out_buffer_size);
    +       if (!_deepcopy_out_buffer)
    +       {
    +           _result = OE_OUT_OF_MEMORY;
    +           goto done;
    +       }
    +   }

    +   /* Serialize the deep-copied content into the buffer. */
    +   if (pargs_in->ptr && pargs_in->ptr[_i_1].blob_array)
    +       OE_WRITE_DEEPCOPY_OUT_PARAM(pargs_in->ptr[_i_1].blob_array, ((size_t)pargs_in->ptr[_i_1].num * sizeof(Blob)));
    +   for (size_t _i_2 = 0; _i_2 < pargs_in->ptr[_i_1].num; _i_2++)
    +   {
    +       if (pargs_in->ptr && pargs_in->ptr[_i_1].blob_array && pargs_in->ptr[_i_1].blob_array[_i_2].buf)
    +           OE_WRITE_DEEPCOPY_OUT_PARAM(pargs_in->ptr[_i_1].blob_array[_i_2].buf, pargs_in->ptr[_i_1].blob_array[_i_2].len);
    +   }

    +   if (_deepcopy_out_buffer_offset != _deepcopy_out_buffer_size)
    +   {
    +       _result = OE_FAILURE;
    +       goto done;
    +   }
    +   OE_ADD_SIZE(output_buffer_offset, _deepcopy_out_buffer_size);

>     +   /* Allocate a host buffer and copy the content over to the buffer. */

Reviewer 4: Instead of this, we could generate:
```
pargs_out->_host_buffer = _deepcopy_out_buffer;
pargs_out->_host_buffer_size = deepcopy_out_buffer_size;
```

The infrastructure can marshal this over to the host as appropriate. The generated code doesn't have to deal with host memory.

Author: Updated.

    +       free(_deepcopy_out_buffer);

    +   /* Free nested buffers allocated by the host. */
    +   if (pargs_in->ptr)
    +   {
    +       for (size_t _i_1 = 0; _i_1 < 1; _i_1++)
    +       {
    +           if (pargs_in->ptr[_i_1].blob_array)
    +           {
    +               for (size_t _i_2 = 0; _i_2 < pargs_in->ptr[_i_1].num; _i_2++)
    +               {
    +                   free(pargs_in->ptr[_i_1].blob_array[_i_2].buf);
    +                   pargs_in->ptr[_i_1].blob_array[_i_2].buf = NULL;
    +               }
    +           }
    +           free(pargs_in->ptr[_i_1].blob_array);
    +           pargs_in->ptr[_i_1].blob_array = NULL;
    +       }
    +   }

        if (output_buffer_size >= sizeof(*pargs_out) &&
            oe_is_within_enclave(pargs_out, output_buffer_size))
            pargs_out->_result = _result;
    }
    ```

- OCALL

  Similar to the case of ECALL, the design adds a `_host_buffer` member to the internal struct that holds the shared pointer.
  The pointer holds the host-set content after `foo` successfully invokes `oe_call_host_function`.
  ```diff
  typedef struct _foo_args_t
  {
      oe_result_t _result;

>   +   uint8_t* _host_buffer;

Reviewer 4: As for ecall, add `_host_buffer_size`.

Author: Updated.

    +   }

    +   if (_deepcopy_out_buffer_size)
    +   {
    +       _deepcopy_out_buffer = (uint8_t*) malloc(_deepcopy_out_buffer_size);
    +       if (!_deepcopy_out_buffer)
    +       {
    +           _result = OE_OUT_OF_MEMORY;
    +           goto done;
    +       }
    +   }

    +   /* Serialize the deep-copied content into the buffer. */
    +   if (pargs_in->ptr && pargs_in->ptr->blob_array)
    +       OE_WRITE_DEEPCOPY_OUT_PARAM(pargs_in->ptr->blob_array, ((size_t)pargs_in->ptr->num * sizeof(Blob)));
    +   for (size_t _i_2 = 0; _i_2 < pargs_in->ptr->num; _i_2++)
    +   {
    +       if (pargs_in->ptr && pargs_in->ptr->blob_array && pargs_in->ptr->blob_array[_i_2].buf)
    +           OE_WRITE_DEEPCOPY_OUT_PARAM(pargs_in->ptr->blob_array[_i_2].buf, pargs_in->ptr->blob_array[_i_2].len);
    +   }

    +   if (_deepcopy_out_buffer_offset != _deepcopy_out_buffer_size)
    +   {
    +       _result = OE_FAILURE;
    +       goto done;
    +   }
    +   OE_ADD_SIZE(output_buffer_offset, _deepcopy_out_buffer_size);

    +   /* Pass the pointer of the deep-copied content to the enclave. */

>     +   pargs_out->_host_buffer = _deepcopy_out_buffer;

Reviewer 4: Once the earlier suggestions for ecall are incorporated, the generated code for ecall/ocall deep-copy out parameters will look identical.

Author: Updated. Currently, the enclave (OCALL wrapper) performs an additional check against the buffer to ensure it resides in enclave memory.

        /* Propagate errno back to enclave. */
        /* Errno propagation not enabled. */

        /* Success. */
        _result = OE_OK;
        *output_bytes_written = output_buffer_offset;

    done:
    +   /* Free pargs_out->_host_buffer on failure. */
    +   if (_result != OE_OK)
    +   {
    +       free(pargs_out->_host_buffer);
    +       pargs_out->_host_buffer = NULL;
    +   }

    +   /* Free nested buffers allocated by the host. */
    +   if (pargs_in->ptr)
    +   {
    +       for (size_t _i_1 = 0; _i_1 < 1; _i_1++)
    +       {
    +           if (pargs_in->ptr[_i_1].blob_array)
    +           {
    +               for (size_t _i_2 = 0; _i_2 < pargs_in->ptr[_i_1].num; _i_2++)
    +               {
    +                   free(pargs_in->ptr[_i_1].blob_array[_i_2].buf);
    +                   pargs_in->ptr[_i_1].blob_array[_i_2].buf = NULL;
    +               }
    +           }
    +           free(pargs_in->ptr[_i_1].blob_array);
    +           pargs_in->ptr[_i_1].blob_array = NULL;
    +       }
    +   }

        if (pargs_out && output_buffer_size >= sizeof(*pargs_out))
            pargs_out->_result = _result;
    }
    ```

Discussion
-----------

> The proposed design relies on the shared memory between an enclave and a host and therefore does not apply

Reviewer 4: By moving host buffer allocation to the infrastructure rather than the generated code, there is no longer any reliance on edger8r generated code on shared memory. It is the infrastructure that needs to marshal the deep-copy parameter appropriately on each platform (SGX, OPTEE).

Author: Removed the section.
